x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsTask.java

```diff
@@ -331,14 +331,26 @@ void persistProgress(Client client, String jobId, Runnable runnable) {
             }
         );
 
-        // Step 1: Search for existing progress document in .ml-state*
-        SearchRequest searchRequest =
-            new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern())
-                .source(
-                    new SearchSourceBuilder()
-                        .size(1)
-                        .query(new IdsQueryBuilder().addIds(progressDocId)));
-        executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, searchFormerProgressDocListener);
+        // Step 2: Search for existing progress document in .ml-state*
+        ActionListener<Void> reindexProgressUpdateListener = ActionListener.wrap(
+            aVoid -> {
+                SearchRequest searchRequest =
+                    new SearchRequest(AnomalyDetectorsIndex.jobStateIndexPattern())
+                        .source(
+                            new SearchSourceBuilder()
+                                .size(1)
+                                .query(new IdsQueryBuilder().addIds(progressDocId)));
+                executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, searchFormerProgressDocListener);
+            },
+            e -> {
+                LOGGER.error(new ParameterizedMessage(
+                    "[{}] cannot persist progress as an error occurred while updating reindexing task progress", taskParams.getId()), e);
+                runnable.run();
+            }
+        );
+
+        // Step 1: Update reindexing progress as it could be stale
+        updateReindexTaskProgress(reindexProgressUpdateListener);
```

Comment:
I was going to comment about "what else could be stale", but looking at how we handle the `_stats` call, we effectively do this.

So, we are just shortcutting the `_stats` call here by slightly "duplicating" the code here 

I am referencing:

https://github.com/elastic/elasticsearch/blob/c5cef17bd597eb42164e518c94308b5093025107/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java#L122

Reply:
Yes indeed. Reindex is the only async task that we need to take care of updating its progress before reporting it. I thought about adding back the call to the get stats API. But it is quite a long way around. What do you think is better?

Reply:
I think doing this directly is fine.

We should definitely comment here AND on the _stats call. That way if ever we introduce another async service that we have to worry about refreshing, we capture it.